DEVELOPMENT.md

```diff
@@ -0,0 +1,229 @@
+# Development
+
+M3 is a very active, large codebase. This guide is intended for all potential
+contributors, designed to make it easy to get started with contributing to M3.
+
+## Setup your development environment
+
+First, fork https://github.com/m3db/m3 to your own user or org account on Github.
+
+Then, clone your fork:
+
+```bash
+mkdir -p $GOPATH/src/github.com/m3db
+cd $GOPATH/src/github.com/m3db
+git clone git@github.com:<fork_user_or_org>/m3.git
+```
+
+Note: All commits/branches/etc must be pushed to the fork, not to m3db/m3.
+Verify that `git clone` properly set the repository's upstream:
+
+```
+# git remote -v
+origin  git@github.com:<fork_user_or_org>/m3.git (fetch)
+origin  git@github.com:<fork_user_or_org>/m3.git (push)
+```
+
+Install dependencies:
+
+```bash
+cd m3/
+make install-vendor-m3
+```
+
+If everything is set up correctly, `m3dbnode` should build successfully:
+
+```bash
+make m3dbnode
+```
+
+## Running the M3 stack locally
+
+Follow the instructions in [this README][local-readme].
+
+[local-readme]: ./scripts/development/m3_stack/README.md
+
+## Updating Mocks And Generated Files
+
+If changes require updates to mocks or other generated files, make sure to
+update those files. There are `make` targets to help with generation:
+
+- Mocks: `make mock-gen`
+- Protobuf: `make proto-gen`
+- Thrift: `make thrift-gen`
+
+Don't forget to account for changes to generated files in tests!
+
+## Testing Changes
+
+M3 has an extensive test suite to ensure that we are able to validate changes.
+More notes about the various testing strategies we employ can be found in
+[TESTING][TESTING.md].
+
+While our CI suite will run all tests prior to allowing pull requests to be
+merged, developers should test their changes during development and prior to
+pushing updates. To test code, simply use `go test`:
+
+```
+go test ./... -v
```

Comment:
```
go: finding module for package github.com/m3db/m3/src/x/generated/proto/test
src/x/test/namespace_extended_opts.go:27:2: no matching versions for query "latest"
```

Is that expected?

Reply:
This path should be `./src/...` instead of `./...`, but otherwise it works for me:

```
# go test ./src/x/...
ok      github.com/m3db/m3/src/x/cache  3.384s
<snip>
# echo $?
0
```

Reply:
@mway 

`go test ./src/... -v` I still get the same error, run from the root directory?

Reply:
I think there might be something wrong with your Go environment, possibly you haven't cloned this within your GOPATH.

Reply:
@ChrisChinchilla Were you able to verify your Go environment?

Reply:
@mway odd, I changed nothing, recloned the repo and this time it worked